### PR TITLE
Implement standalone class generation for multiclass BDTG

### PR DESCRIPTION
**Pull request by biemster**
Standalone C++ class generation was not implemented for multiclass BDTG. This change uses the logic of TMVA::MethodBDT::GetMulticlassValues() and uses it in a new function ReadBDTG::GetMulticlassValues() when the multiclass training is used.
Tested on multiclass BDTG and binary BDTG training, in both cases a compiling and functioning standalone class is generated.

### CODE REVIEW

File: tmva/tmva/src/MethodBDT.cxx [hunk 1]
```diff
@@ -2765,50 +2765,93 @@ void TMVA::MethodBDT::MakeClassSpecific( std::ostream& fout, const TString& clas
...
+      fout << "         if(iClass!=j)" << std::endl;
+      fout << "            norm += temp[j] / temp[iClass];" << std::endl;
+      fout << "      }" << std::endl;
+      fout << "      (*fMulticlassReturnVal).push_back(1.0/(1.0+norm));" << std::endl;
+      fout << "   }" << std::endl;
+      fout << std::endl;
+      fout << "   return *fMulticlassReturnVal;" << std::endl;
+      fout << "}" << std::endl;
+   }
+   else {
```
Comment (ashlaban) [inline tmva/tmva/src/MethodBDT.cxx]: } else {: bracket and else on the same line. Contrary to the impression one might get when reading the code, the closing bracket and the else should be on the same line

File: tmva/tmva/src/MethodBase.cxx [hunk 2]
```diff
@@ -3039,7 +3039,12 @@ void TMVA::MethodBase::MakeClass( const TString& theClassFileName ) const
    fout << "   virtual ~IClassifierReader() {}" << std::endl;
    fout << std::endl;
    fout << "   // return classifier response" << std::endl;
-   fout << "   virtual double GetMvaValue( const std::vector<double>& inputValues ) const = 0;" << std::endl;
+   if(GetAnalysisType() == Types::kMulticlass) {
+      fout << "   virtual std::vector<Float_t>& GetMulticlassValues( const std::vector<double>& inputValues ) const = 0;" << std::endl;
```
Comment (ashlaban) [inline tmva/tmva/src/MethodBase.cxx]: This should return the vector by value. Basically the compiler does return value optimisation for this kind of function so there should in practise be no performance hit on this, and we don't have to worry about managing memory.

File: tmva/tmva/src/MethodBase.cxx [hunk 3]
```diff
@@ -3118,7 +3123,12 @@ void TMVA::MethodBase::MakeClass( const TString& theClassFileName ) const
    fout << "   // the classifier response" << std::endl;
    fout << "   // \"inputValues\" is a vector of input values in the same order as the" << std::endl;
    fout << "   // variables given to the constructor" << std::endl;
-   fout << "   double GetMvaValue( const std::vector<double>& inputValues ) const override;" << std::endl;
+   if(GetAnalysisType() == Types::kMulticlass) {
+      fout << "   std::vector<Float_t>& GetMulticlassValues( const std::vector<double>& inputValues ) const override;" << std::endl;
```
Comment (ashlaban) [inline tmva/tmva/src/MethodBase.cxx]: Return by value (same as above!)

File: tmva/tmva/src/MethodBDT.cxx [hunk 4]
```diff
@@ -2837,6 +2881,9 @@ void TMVA::MethodBDT::MakeClassSpecificHeader(  std::ostream& fout, const TStrin
    TString nodeName = className;
    nodeName.ReplaceAll("Read","");
    nodeName.Append("Node");
+   fout << "#include <algorithm>" << std::endl;
+   fout << "#include <limits>" << std::endl;
+   fout << "   " << std::endl;
```
Comment (ashlaban) [inline tmva/tmva/src/MethodBDT.cxx]: Very minor. This should probably be just `fout << "" << std::endl;`

File: tmva/tmva/src/MethodBDT.cxx [hunk 5]
```diff
@@ -2765,50 +2765,93 @@ void TMVA::MethodBDT::MakeClassSpecific( std::ostream& fout, const TString& clas
...
-         if (fIsHighBkgCut[ivar]){
-            fout << "   if (inputValues["<<ivar<<"] > "<<fHighBkgCut[ivar] <<")  return -1;  // is background preselection cut" << std::endl;
-         }
-         if (fIsHighSigCut[ivar]){
-            fout << "   if (inputValues["<<ivar<<"] > "<<fHighSigCut[ivar]<<")  return  1;  // is signal preselection cut" << std::endl;
+
+   if(GetAnalysisType() == Types::kMulticlass) {
+      fout << "std::vector<Float_t>& ReadBDTG::GetMulticlassValues__( const std::vector<double>& inputValues ) const" << std::endl;
+      fout << "{" << std::endl;
+      fout << "   std::vector<Float_t> *fMulticlassReturnVal = new std::vector<Float_t>();" << std::endl;
```
Comment (ashlaban) [inline tmva/tmva/src/MethodBDT.cxx]: We know the desired size of the vector here, and we know that we will fill every element so I believe the preferred approach is
```
std::vector<Float_t> multiclassReturnVal;
multiclassReturnVal.reserve(nClasses);
```

File: tmva/tmva/src/MethodBDT.cxx (hunk 4 above)
Comment (biemster) [reply tmva/tmva/src/MethodBDT.cxx]: This is actually copy paste old code, i fixed a few more, and i think fout << std::endl; is even better :)

File: tmva/tmva/src/MethodBDT.cxx (hunk 1 above)
Comment (biemster) [reply tmva/tmva/src/MethodBDT.cxx]: I found a few more of those, I probably should have scanned through the coding convention doc a bit..

File: tmva/tmva/src/MethodBase.cxx (hunk 2 above)
Comment (biemster) [reply tmva/tmva/src/MethodBase.cxx]: done

**Comment by ashlaban** [comment]
Results verified! The PR is ready for merging. However, in testing it was discovered that there is a difference in output between `TMVApplication` and stand-alone execution of the BDT's. This is due to a difference in the `GoesRight` function. `TMVA::DecisionTree` cuts on `>=` while the stand-alone cuts on `>`.

@biemster If you could include the fix to `GoesRight()` here, that would be terrific.

**Comment by biemster** [comment]
I'll commit one more patch with a fix to `GoesRight()` tonight or tomorrow to this branch.

**Comment by ashlaban** [comment]
@biemster Thank you! This looks good to me now.
@lmoneta I think this is ready to merge now!

**Comment by ashlaban** [comment]
Thanks for you hard work @biemster!

**Comment by biemster** [comment]
No problem, happy to help and same to you and all ROOTers!

**Pull request merged**